src/checkers/inference/solver/backend/encoder/SlotSlotCombo.java

```diff
@@ -75,6 +76,20 @@ public static SlotSlotCombo valueOf(CombineConstraint combineConstraint) {
         return valueOf(combineConstraint.getTarget(), combineConstraint.getDeclared());
     }
 
+    /**
+     * Gets the {@code SlotSlotCombo} of a {@link ArithmeticConstraint}
+     *
+     * @param arithmeticConstraint {@code ArithmeticConstraint} whose {@code SlotSlotCombo} is
+     *        analyzed
+     * @return {@code SlotSlotCombo} of the passed-in {@code arithmeticConstraint}
+     *
+     * @see ArithmeticConstraint
+     */
+    public static SlotSlotCombo valueOf(ArithmeticConstraint arithmeticConstraint) {
```

Comment:
Adding this method makes `SlotSlotCombo` be aware of the detail of `ArithmeticConstraint`.

What is the motivation of this method? Is this method be used multi-places and you need to share it?

I only saw it in `ConstraintEncoderCoordinator#dispatch(ArithmeticConstraint)`, if that is the only place,
maybe remove this coupling would be better.

Reply:
I can do that, though, shouldn't we also get rid of it for VPA/Combine constraint? @topnessman

Reply:
I don't think this "coupling" is a problem. It's just a util method that gets the slot combo of two ingredient slots. All the logic to get the type of slotslotcombo is centralized in this class. It seems an overkill to have another class to do the gluing.

Reply:
Every instance of a constraint (with >= 2 slots) should be able to know by themselves whether they are CC/CV/VC/VV. Maybe we can refactor the constraint-specific `valueOf( xxxConstraint )` methods and put them inside the corresponding constraint classes. Even then, the `SlotSlotCombo` class should stay as it contains the shared logic aside from which specific slot is used as the first or second slots in the combo.

If we think the refactoring makes it a better design, lets do it in a separate PR.

Reply:
> I don't think this "coupling" is a problem. It's just a util method that gets the slot combo of two ingredient slots.

It indeed increased unnecessary coupling, Mier. It makes we need to remember one more place to change if we want to change something in these constraint classes. Put it just to the specific constraint `dispatch` method is easier to remember.

Also, it seems these `valueOf(Constraint)` methods just be used in `dispatch` method, so why doing this? It adds more code and just one-line wrapper for one place use. It actually add  more maintenance burden and does save anything actually.

@jyluo Jeff, it should be an easy and small change. At least if we agree on this, you should not add another one-line wrapper method in this PR. And if you remove that wrapper, it is easy to remove the existed wrapper method by the way.
 So, if we agree on this, I think we can directly clean this in this PR.

Reply:
PR for removing these wrapper methods:
https://github.com/opprop/checker-framework-inference/pull/158/files

Reply:
#158 looks good to me. I'll fix the arithmetic valueof in this PR after we merge 158.